paparazzi/src/main/java/app/cash/paparazzi/internal/apng/ApngWriter.kt

```diff
@@ -0,0 +1,277 @@
+import java.io.Closeable
+import java.io.File
+import java.lang.IllegalStateException
+import java.lang.Integer.max
+import java.util.zip.CRC32
+import java.util.zip.Deflater
+
+/**
+ * Provides a basic implementation for encoding APNG's.
+ *
+ * Refs: https://www.w3.org/TR/png/#4Concepts.Encoding, https://wiki.mozilla.org/APNG_Specification
+ */
+internal class ApngWriter(
+  private val path: Path,
+  private val totalFrameCount: Int,
+  fps: Int,
+  fileSystem: FileSystem = FileSystem.SYSTEM
+) : Closeable {
+
+  private val fpsNum = 1
+  private val fpsDen = fps
+  private val tempFile = File(FileSystem.SYSTEM_TEMPORARY_DIRECTORY.toFile(), "temp-${path.name}").apply { deleteOnExit() }
+  private val fileHandle = fileSystem.openReadWrite(tempFile.path.toPath())
+  private val sink = fileHandle.sink().buffer()
+  private val crcEngine = CRC32()
+
+  private lateinit var firstFrame: BufferedImage
+  private lateinit var previousFrame: BufferedImage
+
+  private var maxWidth = 0
+  private var maxHeight = 0
+  private var ihdrWidth = 0
+  private var ihdrHeight = 0
+  private var iDATChunkEndOffset = 0L
+  private var frameCount = 0
+  private var sequenceNumber = 0
+
+  fun writeImage(image: BufferedImage) {
+    if (frameCount == 0) {
+      sink.writePngSignature()
+      sink.writeIHDR(image)
+
+      if (totalFrameCount > 1) {
+        sink.writeACTL(totalFrameCount, 0)
+        sink.writeFCTL(sequenceNumber++, Rectangle(image.width, image.height))
+      }
+
+      sink.writeIDAT(image)
+      iDATChunkEndOffset = fileHandle.position(sink)
+      firstFrame = image
+    } else {
+      val subRect = ImageUtils.smallestDiffRect(previousFrame, image) ?: MIN_FRAME_RECT
+      sink.writeFCTL(sequenceNumber++, subRect)
+      sink.writeFDAT(sequenceNumber++, subRect, image)
+    }
+
+    maxWidth = max(maxWidth, image.width)
+    maxHeight = max(maxHeight, image.height)
+    previousFrame = image
+    frameCount++
+  }
+
+  override fun close() {
+    if (frameCount != totalFrameCount) {
+      throw IllegalStateException("Expected $totalFrameCount total frames, actual count $frameCount")
+    }
+
+    sink.writeIEND()
+    fileHandle.resize(fileHandle.position(sink))
+    sink.close()
+
+    if (ihdrWidth == maxWidth && ihdrHeight == maxHeight) {
+      tempFile.copyTo(path.toFile(), true)
+    } else {
+      correctFirstFrame()
+    }
+
+    fileHandle.close()
+  }
+
+  /**
+   * The IHDR chunk defines the image size from which all animation frames must be
+   * encapsulated. If over the course of the animation we need to render a larger
+   * image than the first frame we need to resize the first frame such that the
+   * entire animation will fit within. Subsequent animation frames are still correct
+   * and don't need to be re-computed.
+   */
+  private fun correctFirstFrame() {
+    val resizedFirstFrame = firstFrame.resize(maxWidth, maxHeight)
+    val stableHandle = FileSystem.SYSTEM.openReadWrite(path)
```

Comment:
```suggestion
    val stableHandle = fileSystem.openReadWrite(path)
```
Also needs the ctor parameter to become a `private val`